### PR TITLE
Get by ID and Get All endpoints

### PR DESCRIPTION
**Pull request by notatiyyah**
## Describe this PR

Create the 'Get All Technologies' endpoint and 'Get Technology by ID' endpoint

#### _Checklist_

- [x] Added relevant unit and end-to-end tests
- [x] Added tests to cover all new production code
- [x] Added a README
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: TechRadarApi/V1/Gateways/TechnologyGateway.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+        public TechnologyGateway(IDynamoDBContext dynamoDbContext)
+        {
+            _dynamoDbContext = dynamoDbContext;
+        }
+
+        public List<Technology> GetAll()
+        {
+            var conditions = new List<ScanCondition>();
+            var results = _dynamoDbContext.ScanAsync<TechnologyDbEntity>(conditions).GetRemainingAsync().GetAwaiter().GetResult();
```
Comment (humulla) [inline TechRadarApi/V1/Gateways/TechnologyGateway.cs]: The GetAwaiter() and GetResult() doesn't actually do anything, please read this:
[https://gsferreira.com/archive/2020/08/avoid-getawaiter-getresult-at-all-cost/](https://gsferreira.com/archive/2020/08/avoid-getawaiter-getresult-at-all-cost/ )

Change the method to be an async instead

File: TechRadarApi/V1/Controllers/TechRadarApiController.cs [hunk 2]
```diff
@@ -2,50 +2,43 @@
...
-        [ProducesResponseType(typeof(ResponseObjectList), StatusCodes.Status200OK)]
+        [ProducesResponseType(typeof(TechnologyResponseObjectList), StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
         [HttpGet]
-        public IActionResult ListContacts()
+        public IActionResult ListTechnologies()
         {
-            return Ok(_getAllUseCase.Execute());
+            var result = _getAllUseCase.Execute();
+            if (result.Technologies.Count == 0) return NoContent();
```
Comment (humulla) [inline TechRadarApi/V1/Controllers/TechRadarApiController.cs]: I dont think the response should be 204, change the method to be a try catch and in the catch we can have a 400 response with a message that there are no technologies

File: TechRadarApi.Tests/V1/Controllers/TechRadarControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            actualResponse.StatusCode.Should().Be(200);
+            actualResponse.Value.Should().BeEquivalentTo(expectedResponse);
+        }
+
+        [Test]
+        public void GetTechnologyWithNonExistentIDReturnsNotFoundResponse()
+        {
+            // Arrange
+            var id = Guid.NewGuid();
+            // Act
```
Comment (humulla) [inline TechRadarApi.Tests/V1/Controllers/TechRadarControllerTests.cs:52]: Add the SetUp() to the usecase here where the return would be:
`Returns((TechnologyResponseObject) null)`

File: TechRadarApi.Tests/V1/UseCase/GetAllTechnologiesUseCaseTests.cs [hunk 4]
```diff
@@ -13,27 +13,29 @@ namespace TechRadarApi.Tests.V1.UseCase
...
-
-            _classUnderTest.Execute().Should().BeEquivalentTo(expectedResponse);
+            var expectedResponse = new TechnologyResponseObjectList { Technologies = stubbedEntities.ToResponse() };
+            // Act
+            var actualResponse = _classUnderTest.Execute();
+            // Assert
+            actualResponse.Should().BeEquivalentTo(expectedResponse);
         }
 
         //TODO: Add extra tests here for extra functionality added to the use case
```
Comment (humulla) [inline TechRadarApi.Tests/V1/UseCase/GetAllTechnologiesUseCaseTests.cs]: Add extra test for when the api responds with a 400

File: TechRadarApi.Tests/V1/UseCase/GetTechnologyByIdUseCaseTests.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        public void GetsTechnologyFromGatewayUsingId()
+        {
+            // Arrange
+            var stubbedEntity = _fixture.Create<Technology>();
+            _mockGateway.Setup(x => x.GetTechnologyById(stubbedEntity.Id)).Returns(stubbedEntity);
+            // Act
+            var response = _classUnderTest.Execute(stubbedEntity.Id);
+            // Assert
+            response.Should().BeEquivalentTo(stubbedEntity);
+        }
```
Comment (humulla) [inline TechRadarApi.Tests/V1/UseCase/GetTechnologyByIdUseCaseTests.cs:64]: Add another test for 404 response

File: TechRadarApi.Tests/V1/UseCase/GetTechnologyByIdUseCaseTests.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    }
```
Comment (humulla) [inline TechRadarApi.Tests/V1/UseCase/GetTechnologyByIdUseCaseTests.cs:65]: Add another test for 500 response

File: TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+using System;
+using System.Collections.Generic;
+
+namespace TechRadarApi.Tests.V1.Gateways
+{
+    [TestFixture]
+    public class TechnologyGatewayTests
+    {
+        private readonly Fixture _fixture = new Fixture();
+        private Mock<IDynamoDBContext> _dynamoDb;
```
Comment (humulla) [inline TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs]: Don't mock the DynamodbContext, instead inject the integrationTest class and use the dynamodbContext from there (the integration test class creates the table as well)

File: TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        [Test]
+        public void GetTechnologyByIdReturnsTheTechnologyIfItExists()
+        {
+            // Arrange
+            var entity = _fixture.Create<Technology>();
+            var dbEntity = DatabaseEntityHelper.CreateDatabaseEntityFrom(entity);
+
+            _dynamoDb.Setup(x => x.LoadAsync<TechnologyDbEntity>(entity.Id.ToString(), default))
```
Comment (humulla) [inline TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs]: You need to use SaveAsync() to save the data to the database

File: TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                     .ReturnsAsync(dbEntity);
+
+            // Act
+            var response = _classUnderTest.GetTechnologyById(entity.Id);
+
+            // Assert
+            _dynamoDb.Verify(x => x.LoadAsync<TechnologyDbEntity>(entity.Id.ToString(), default), Times.Once);
+
+            response.Id.Should().Be(entity.Id.ToString());
```
Comment (humulla) [inline TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs]: You could change this to something like  `response.should().BeEquivalent(entity)`

File: TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+            List<ScanCondition> conditions = new List<ScanCondition>();
+            _dynamoDb.Setup(x => x.ScanAsync<TechnologyDbEntity>(conditions, default).GetRemainingAsync(default))
+                    .ReturnsAsync(stubbedResponse);
+            // Act
+            var response = _classUnderTest.GetAll();
+            // Assert
+            _dynamoDb.Verify(x => x.ScanAsync<TechnologyDbEntity>(conditions, default).GetRemainingAsync(default), Times.Once);
+            response.Should().BeEquivalentTo(entities);
+        }
```
Comment (humulla) [inline TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs:119]: Add a 500 exception test as well

File: TechRadarApi.Tests/V1/E2ETests/GetByIdEndToEndTests.cs [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+            // Assert
+            response.StatusCode.Should().Be(HttpStatusCode.OK);
+            apiEntity.Should().BeEquivalentTo(entity);
+        }
+
+        [Test]
+        public async Task GetTechnologyByInvalidIdReturnsBadRequestResponse()
+        {
+            // Arrange
+            var uri = new Uri($"api/v1/technologies/1", UriKind.Relative);
```
Comment (humulla) [inline TechRadarApi.Tests/V1/E2ETests/GetByIdEndToEndTests.cs]: Dont hardcode the id value

File: TechRadarApi.Tests/V1/E2ETests/GetAllEndToEndTests.cs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+        [Test]
+        public async Task GetAllTechnologiesReturnsNoContentResponseWhenTheTableIsEmpty()
+        {
+            // Arrange
+            var uri = new Uri($"api/v1/technologies", UriKind.Relative);
+            // Act
+            var response = await Client.GetAsync(uri).ConfigureAwait(false);
+            // Assert
+            response.StatusCode.Should().Be(HttpStatusCode.NoContent);
```
Comment (humulla) [inline TechRadarApi.Tests/V1/E2ETests/GetAllEndToEndTests.cs]: The status code may change to 400 if the controller logic works

**Review by humulla — CHANGES_REQUESTED** [review]
Great start! There are some changes that need to be made as commented, please also change all tests to use the Xunit framework. Let me know if you have any questions

File: TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs [hunk 13]
```diff
@@ -39,90 +68,70 @@ public async Task GetTechnologyByIdReturnsNullIfTechnologyDoesntExist()
...
             var response = await _classUnderTest.GetAll().ConfigureAwait(false);
             // Assert
-            _dynamoDb.Verify(x => x.ScanAsync<TechnologyDbEntity>(conditions, default).GetRemainingAsync(default), Times.Once);
             response.Should().BeEquivalentTo(entities);
         }
 
-        [Test]
-        [Ignore("Getting a bug - can't stub the DB response")]
-
+        [Fact(Skip = "Cannot mock DynamoDb context")]
```
Comment (humulla) [inline TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs]: Why is this test skipped? what is the issue you are facing

File: TechRadarApi/V1/Controllers/TechRadarApiController.cs [hunk 14]
```diff
@@ -2,50 +2,44 @@
...
+            return Ok(result);
         }
 
-        /// <summary>
-        /// ...
-        /// </summary>
-        /// <response code="200">...</response>
-        /// <response code="404">No ? found for the specified ID</response>
-        [ProducesResponseType(typeof(ResponseObject), StatusCodes.Status200OK)]
+        [ProducesResponseType(typeof(TechnologyResponseObject), StatusCodes.Status200OK)]
```
Comment (humulla) [inline TechRadarApi/V1/Controllers/TechRadarApiController.cs:33]: add a response object for status code 404 as well here

File: TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs (hunk 13 above)
Comment (notatiyyah) [reply TechRadarApi.Tests/V1/Gateways/TechnologyGatewayTests.cs]:
```
mockDynamoDb.Setup(x => x.ScanAsync<TechnologyDbEntity>(conditions, default).GetRemainingAsync(default))
                     .ThrowsAsync(exception);
```
This line ^ doesn't work with the scan async method as it's a 'non-virtual method'. Here's [a link](https://github.com/moq/moq4/issues/590) for more info. Should I just remove the test?

File: TechRadarApi/V1/Controllers/TechRadarApiController.cs (hunk 14 above)
Comment (notatiyyah) [reply TechRadarApi/V1/Controllers/TechRadarApiController.cs:33]: What would that be for? If the table is empty, should return an empty list and 200 OK.

Comment (humulla) [reply TechRadarApi/V1/Controllers/TechRadarApiController.cs:33]: No it would be if the id they provide is not available in the table

Comment (notatiyyah) [reply TechRadarApi/V1/Controllers/TechRadarApiController.cs:33]: Oops, didn't realise what endpoint I was looking at 😅

File: TechRadarApi.Tests/V1/E2ETests/GetAllEndToEndTests.cs (hunk 12 above)
Comment (humulla) [inline TechRadarApi.Tests/V1/E2ETests/GetAllEndToEndTests.cs]: This should change to return an 200 response

File: TechRadarApi/V1/Controllers/TechRadarApiController.cs [hunk 15]
```diff
@@ -2,50 +2,43 @@
...
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
```
Comment (humulla) [inline TechRadarApi/V1/Controllers/TechRadarApiController.cs]: This should be removed as we are no longer using this

**Review by humulla — APPROVED** [review]
Looks good to me!

**Pull request merged**